consensus/polybft/validators_snapshot.go

```diff
@@ -214,29 +221,96 @@ func (v *validatorsSnapshotCache) storeSnapshot(block uint64, snapshot AccountSe
 // Cleanup cleans the validators cache in memory and db
 func (v *validatorsSnapshotCache) cleanup() error {
 	if len(v.snapshots) >= validatorSnapshotLimit {
-		latestBlock := uint64(0)
+		latestEpoch := uint64(0)
 
 		for e := range v.snapshots {
-			if e > latestBlock {
-				latestBlock = e
+			if e > latestEpoch {
+				latestEpoch = e
 			}
 		}
 
-		block := latestBlock
-		cache := make(map[uint64]AccountSet, numberOfSnapshotsToLeaveInMemory)
+		startEpoch := latestEpoch
+		cache := make(map[uint64]*validatorSnapshot, numberOfSnapshotsToLeaveInMemory)
 
 		for i := 0; i < numberOfSnapshotsToLeaveInMemory; i++ {
-			if snapshot, exists := v.snapshots[block]; exists {
-				cache[block] = snapshot
+			if snapshot, exists := v.snapshots[startEpoch]; exists {
+				cache[startEpoch] = snapshot
 			}
 
-			block--
+			startEpoch--
 		}
 
 		v.snapshots = cache
 
-		return v.state.cleanValidatorSnapshotsFromDB(latestBlock)
+		return v.state.cleanValidatorSnapshotsFromDB(latestEpoch)
 	}
 
 	return nil
 }
+
+// getLastCachedSnapshot gets the latest snapshot cached
+// If it doesn't have snapshot cached for desired epoch, it will return the latest one it has
+func (v *validatorsSnapshotCache) getLastCachedSnapshot(currentEpoch uint64) (*validatorSnapshot, error) {
+	snapshot := v.snapshots[currentEpoch]
+	if snapshot != nil {
+		v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpoch)
+
+		return snapshot, nil
+	}
+
+	snapshot, err := v.state.getLastSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	return snapshot, nil
+}
```

Comment:
WDYT about this change? Basically idea is to return the more recent snapshot (either it is fetched from the cache or db). Although it makes code slightly complex, I think it is worth considering it.

```suggestion
func (v *validatorsSnapshotCache) getLastCachedSnapshot(currentEpoch uint64) (*validatorSnapshot, error) {
	var (
		cachedSnapshot *validatorSnapshot
	)

	for epoch := currentEpoch; epoch >= 0; epoch-- {
		cachedSnapshot = v.snapshots[epoch]
		if cachedSnapshot != nil {
			v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpoch)

			break
		}
	}

	dbSnapshot, err := v.state.getLastSnapshot()
	if err != nil {
		return nil, err
	}

	if dbSnapshot != nil && cachedSnapshot != nil &&
		dbSnapshot.Epoch > cachedSnapshot.Epoch {
		return dbSnapshot, nil
	}

	return cachedSnapshot, nil
}
```

Reply:
It makes sense.

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/997/commits/a4709904d8b5cfc71dab457a4bf18bbb1338367b